text/vm-2.0.md

```diff
@@ -0,0 +1,228 @@
+    //           I::Item: (Key, ToJs);
+}
+```
+
+**vm.lock()**
+Freezes the VM by taking out a lock.
+
+**vm.argument(i)**
+Return the *i*'th argument of the current function call.
+
+**vm.this()**
+Return the `this` binding of the current function call.
+
+**vm.callee()**
+Return the callee (i.e., the called function) of the current function call.
+
+**vm.execute(|mut vm| { … })**
+Execute a callback with a new `Vm` instance. Any locally allocated handles created during the body of the callback are only kept alive for the duration of the callback. This method is useful for executing code that may generate temporary data and allowing that temporary data to get freed by the JavaScript garbage collector, especially inside of a loop that may result in large amounts of temporary garbage.
+
+**vm.evaluate(|mut vm| { … })**
+Execute a callback with a new `Vm` instance and return a JavaScript value. Any locally allocated handles created during the body of the callback are only kept alive for the duration of the callback, except for the result value, which is kept alive for the duration of the outer `Vm`. This method is useful for executing code that may generate temporary data and allowing that temporary data to get freed by the JavaScript garbage collector, especially inside of a loop that may result in large amounts of temporary garbage.
+
+**vm.number(f)**
+**vm.boolean(b)**
+**vm.string(s)**
+**vm.null()**
+**vm.undefined()**
+Convenience constructors for primitive value types.
+
+**vm.empty_object()**
+**vm.empty_array()**
+Convenience constructors for vanilla objects and array objects.
+
+**vm.array_buffer(byte_len)**
+Convenience constructor for creating a new `ArrayBuffer` object.
+
+## Implementations of `Vm`
+
+The `CallContext`, `ExecuteContext`, and `EvaluateContext` types are the three implementations of `Vm`:
+
+```rust
+struct CallContext<'a, T: This> { /* ... */ };
+
+impl<'a, T: This> Vm<'a, T> for CallContext<'a, T> { /* ... */ };
+```
+
+```rust
+struct ExecuteContext<'a, T: This> { /* ... */ };
+
+impl<'a, T: This> Vm<'a, T> for ExecuteContext<'a, T> { /* ... */ };
+```
+
+```rust
+struct EvaluateContext<'a, T: This> { /* ... */ };
+
+impl<'a, T: This> Vm<'a, T> for EvaluateContext<'a, T> { /* ... */ };
+```
+
+These types correspond to different context-sensitive views of the JavaScript virtual machine: the context of a call into a native Rust implementation of a JavaScript function, the context of a call to the `vm.execute()` method, or the context of a call to the `vm.evaluate()` method, respectively. From the user’s perspective, though, the differences between these types are unimportant, and they can simply all be thought of as instances of `Vm`.
+
+## The `Lock` type
+
+The `Lock` type is produced by the `vm.lock()` method described above, and can be used to inspect the internals of JavaScript values:
+
+```rust
+impl Lock {
+    fn inspect<T: Inspect>(&self, value: Handle<T>) -> &mut T::Internals;
```

Comment:
This does not look quite right: the `Handle` is copy, so we can get two aliased `&mut T::Internals`?

Reply:
Yes, good point. The way the current locking mechanism works is by dynamically tracking what ranges in memory are currently being inspected. We can store that state inside the `Lock` object itself, and either make `inspect` an `&mut self` method or store that state in a cell.

Reply:
>The way the current locking mechanism works is by dynamically tracking what ranges in memory are currently being inspected.

Yeah, so I would expect the return type to be `RefMut<T::Internals>`, like `RefCell` works.

Reply:
Really interesting point. If we're going all the way to consistency with the stdlib, it makes me wonder if that suggests refactoring the API so that `Inspect` gets a name more like `RefValue` or something like that, with:
```rust
trait RefValue<T> {
    fn borrow(&self, lock: &Lock) -> Ref<T>;
    fn borrow_mut(&mut self, lock: &Lock) -> RefMut<T>;
}
```
So that you would write examples like `zero_first_byte` from the RFC like so:
```rust
fn zero_first_byte(mut vm: impl Vm) -> JsResult<Handle<JsNumber>> {
    let array_buffer = vm.argument(0).unwrap().check::<JsArrayBuffer>()?;
    let lock = vm.lock();
    let mut contents = array_buffer.borrow_mut(&lock);
    let mut buf = contents.as_u8_slice();
    buf[0] = 0;
    Ok(())
}
```
It's appealing to use the nomenclature of borrowing, although since I haven't worked with ref cells in practice, I don't know what kind of learnability hurdles they might introduce. Maybe it's no worse than the learnability of the `Inspect` API though? I really like reusing the intuitions of borrowing for understanding what it means to temporarily inspect the contents of a JS value.

Reply:
> If we're going all the way to consistency with the stdlib, 

My point was not about consistency with stdlib (though that'd be nice!). I think that if you do run-time tracking, you'll have to use either closure-base APIs, or return some sort of lock-object, like `RefMut`. Returning plain `&mut T` I think will be unsond.

Reply:
@matklad Is there any particular _behavioral_ reason to use `RefMut` instead of `&mut`, or is it just to respect the conventions of the stdlib and send a signal that this is a reference that may dynamically fail to be produced?

Reply:
> Is there any particular behavioral reason to use RefMut instead of &mut, or is it just to respect the conventions of the stdlib and send a signal that this is a reference that may dynamically fail to be produced?

I believe there is a behavioral reason. Specifically, I believe that `&mut` returning API will be unsound. I am not 100% sure about this, but looks like what we are doing here is pretty similar to `RefCell`, and it needs those `RefMut` wrappers to be sound.

Reply:
Hm, I don't see how to make it work with the actual `Ref`/`RefMut` types of `std::cell`, since those have no constructors and the only way to get one is to call `borrow{_mut}()` on a `RefCell`. I could do something analogous with a custom Neon `Ref` and `RefMut` type though.

Reply:
I suppose the soundness issue is needing to track the duration of the borrows? And having a wrapper type allows us to register the borrow on allocation of the Ref{Mut} and unregister it on drop of the Ref{Mut}?

Reply:
Yeah, I totally didn't mean that we should use `std::cell::RefMut`, that's impossible. But we quite probably will have to roll our own `neon::RefMut`.

Reply:
Got it, thanks! This makes sense to me, and I really like the way the API feels.

Reply:
>I suppose the soundness issue is needing to track the duration of the borrows? And having a wrapper type allows us to register the borrow on allocation of the Ref{Mut} and unregister it on drop of the Ref{Mut}?

Exactly, std::RefCell is basically a single-threaded std::Mutex, and std::RefMut is std::LockGuard.